include/aws/mqtt/private/v5/mqtt3_to_mqtt5_adapter_impl.h

```diff
@@ -0,0 +1,285 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#ifndef AWS_MQTT_MQTT3_TO_MQTT5_ADAPTER_IMPL_H
+#define AWS_MQTT_MQTT3_TO_MQTT5_ADAPTER_IMPL_H
+
+#include <aws/mqtt/mqtt.h>
+
+#include <aws/common/mutex.h>
+#include <aws/common/rw_lock.h>
+#include <aws/common/task_scheduler.h>
+#include <aws/mqtt/client.h>
+#include <aws/mqtt/private/client_impl_shared.h>
+#include <aws/mqtt/v5/mqtt5_client.h>
+
+struct aws_mqtt3_to_mqtt5_adapter_operation_base;
+
+struct aws_mqtt3_to_mqtt5_adapter_operation_vtable {
+    void (*destroy_fn)(struct aws_mqtt3_to_mqtt5_adapter_operation_base *operation);
+    void (*fail_fn)(struct aws_mqtt3_to_mqtt5_adapter_operation_base *operation, int error_code);
+    void (*complete_fn)(struct aws_mqtt3_to_mqtt5_adapter_operation_base *operation, void *completion_data);
+};
+
+struct aws_mqtt3_to_mqtt5_adapter_publish_options {
+    struct aws_mqtt_client_connection_5_impl *adapter;
+
+    const struct aws_byte_cursor topic;
+    enum aws_mqtt_qos qos;
+    bool retain;
+    const struct aws_byte_cursor payload;
+
+    aws_mqtt_op_complete_fn *on_complete;
+    void *on_complete_userdata;
+};
+
+enum aws_mqtt3_to_mqtt5_adapter_operation_type {
+    AWS_MQTT3TO5_AOT_PUBLISH,
+    AWS_MQTT3TO5_AOT_SUBSCRIBE,
+    AWS_MQTT3TO5_AOT_UNSUBSCRIBE,
+};
+
+struct aws_mqtt3_to_mqtt5_adapter_operation_base {
+    struct aws_allocator *allocator;
+    const struct aws_mqtt3_to_mqtt5_adapter_operation_vtable *vtable;
+    void *impl;
+
+    /*
+     * Holds an internal reference to the adapter while traveling to the event loop.  Reference gets released
+     * after intake on the event loop.
+     *
+     * We avoid calling back into a deleted adapter by zeroing out the
+     * mqtt5 operation callbacks for everything we've submitted before final mqtt5 client release.
+     */
+    struct aws_mqtt_client_connection_5_impl *adapter;
+    bool holding_adapter_ref;
+
+    struct aws_task submission_task;
+
+    enum aws_mqtt3_to_mqtt5_adapter_operation_type type;
+    uint16_t id;
+};
+
+struct aws_mqtt3_to_mqtt5_adapter_operation_publish {
+    struct aws_mqtt3_to_mqtt5_adapter_operation_base base;
+
+    /*
+     * holds a reference to the MQTT5 client publish operation until the operation completes or our adapter
+     * goes away.
+     *
+     * In the case where we're going away, we zero out the MQTT5 operation callbacks to prevent crash-triggering
+     * notifications.
+     */
+    struct aws_mqtt5_operation_publish *publish_op;
+
+    aws_mqtt_op_complete_fn *on_publish_complete;
+    void *on_publish_complete_user_data;
+};
+
+/*
+
+  Sequencing (PUBLISH example):
+
+  Mqtt311 public API call
+     Create cross thread task
+     Create adapter op -> Create and attach mqtt5 op
+     allocate id and add operation to adapter table
+     submit cross thread task to event loop
```

Comment:
The example is tracking internal ref to adapter so maybe add just above this that it's where we acquire an internal ref to the adapter?

Reply:
Sure